# Applied Finance with Python - Pandas Exercises
## Apple Stock Analysis (2021-2024)
## STUDENT VERSION

**Instructor:** Fatih Kuruscu

This notebook contains exercises to practice Pandas fundamentals using Apple (AAPL) stock data. Work through these exercises independently after the Microsoft example.

**Instructions:**
- Complete all tasks in order
- Write your code in the provided cells
- Test your code to ensure it runs without errors
- Points are indicated for each task
- Total points: 100 (plus 10 bonus points)

In [1]:
import pandas as pd

# Exercise 1: Data Loading and Exploration

## Task 1.1: Load the Data

Load the Apple stock data from `AAPL_2021_2024.csv` into a DataFrame called `df`.

Print the shape (number of rows and columns) of the dataset.

In [37]:
path_to_data = "../../data/AAPL_2021_2024.csv"
df = pd.read_csv(path_to_data)
print(df.shape)

(1004, 8)


## Task 1.2: Initial Exploration

Display:
- The first 10 rows
- The last 5 rows
- A random sample of 8 rows

In [39]:
# df.head(10) #first 10

In [41]:
# df.tail(5)

In [47]:
# df.sample(8)

## Task 1.3: Data Structure

1. Display the DataFrame info using `.info()`
2. Print all column names
3. Check the data type of each column

In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1004 entries, 0 to 1003
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Date          1004 non-null   object 
 1   Open          1004 non-null   float64
 2   High          1004 non-null   float64
 3   Low           1004 non-null   float64
 4   Close         1004 non-null   float64
 5   Volume        1004 non-null   int64  
 6   Dividends     1004 non-null   float64
 7   Stock Splits  1004 non-null   float64
dtypes: float64(6), int64(1), object(1)
memory usage: 62.9+ KB


In [8]:
df.columns

Index(['Date', 'Open', 'High', 'Low', 'Close', 'Volume', 'Dividends',
       'Stock Splits'],
      dtype='object')

In [7]:
df.dtypes

Date             object
Open            float64
High            float64
Low             float64
Close           float64
Volume            int64
Dividends       float64
Stock Splits    float64
dtype: object

## Task 1.4: Summary Statistics

Calculate and display:
1. Descriptive statistics for all numerical columns using `.describe()`
2. The mean, median, min, and max of the `Close` column
3. The total number of missing values in the entire DataFrame

In [9]:
df.describe()

,Open,High,Low,Close,Volume,Dividends,Stock Splits
count,1004.000000,1004.000000,1004.000000,1004.000000,1.004000e+03,1004.000000,1004.0
mean,166.448988,168.213099,164.845332,166.625568,7.374007e+07,0.003700,0.0
std,31.207064,31.331018,31.140717,31.294908,3.015727e+07,0.029135,0.0
min,116.155664,117.492572,113.403769,113.550148,2.323470e+07,0.000000,0.0
25%,142.970364,144.798221,141.613347,143.151909,5.224162e+07,0.000000,0.0
50%,163.907390,165.784778,162.361188,163.541389,6.803675e+07,0.000000,0.0
75%,183.480134,184.941332,181.914835,183.336956,8.860770e+07,0.000000,0.0
max,257.276679,259.179926,256.718662,258.103729,3.186799e+08,0.250000,0.0


In [51]:
df['Close'].max() #min, max

np.float64(258.1037292480469)

In [54]:
df.isnull().sum().sum()

np.int64(0)

## Task 1.5: Data Preparation

1. Convert the `Date` column to datetime format
2. Set the `Date` column as the index (save as `df_indexed`)
3. Display the first 5 rows of `df_indexed`

In [22]:
df['Date'] = pd.to_datetime(df['Date'], utc=True)
df_indexed = df.set_index('Date')
df_indexed.head()

,Open,High,Low,Close,Volume,Dividends,Stock Splits
Date,,,,,,,
2021-01-04 05:00:00+00:00,130.101340,130.189033,123.514422,126.096573,143301900,0.0,0.0
2021-01-05 05:00:00+00:00,125.589887,128.366921,125.141658,127.655602,97664900,0.0,0.0
2021-01-06 05:00:00+00:00,124.449870,127.694610,123.144175,123.358543,155088000,0.0,0.0
2021-01-07 05:00:00+00:00,125.073496,128.259775,124.586297,127.567947,109578200,0.0,0.0
2021-01-08 05:00:00+00:00,129.039251,129.234143,126.895583,128.668991,105158200,0.0,0.0


# Exercise 2: Data Selection and Filtering

## Task 2.1: Column Selection

1. Select only the `Close` column and display the first 10 values
2. Select the columns `Open`, `High`, `Low`, and `Close` (OHLC data) and display the first 10 rows

In [101]:
# df['Close'].head(10)

0    126.096573
1    127.655602
2    123.358543
3    127.567947
4    128.668991
5    125.677559
6    125.502213
7    127.538719
8    125.609406
9    123.884697
Name: Close, dtype: float64

In [103]:
# df[['Open','High','Low','Close']].head(10)

## Task 2.2: Row Slicing

Using slicing notation, extract:
1. Rows 50 to 60
2. The last 20 rows
3. Every 100th row (show Date and Close columns only)

In [107]:
# df[50:61]

In [108]:
# df[-20:]

In [110]:
# df[['Date','Close']][::100]

## Task 2.3: Boolean Filtering - Single Condition

Filter the data to find:
1. All days where the closing price was above $170
2. All days where the volume was greater than 100 million shares

For each filter, display:
- The number of days that meet the condition
- The percentage of total trading days
- The first 5 rows

In [111]:
def task_printer(df_filtered):
    print('#'*75)
    print(f'Total Number of Days that meet the condition: {len(df_filtered)}')
    print(f'Percentage: {round(len(df_filtered) / len(df) * 100,2)}%')
    print(df_filtered.head(5))
    print('#'*75)

In [112]:
df_high_prices = df[df['Close'] > 170]
df_high_shares = df[df['Volume'] > 100_000_000]
task_printer(df_high_prices)
task_printer(df_high_shares)

###########################################################################
Total Number of Days that meet the condition: 415
Percentage: 41.33%
                          Date        Open        High         Low  \
235  2021-12-08 00:00:00-05:00  168.756755  172.511700  167.354772   
236  2021-12-09 00:00:00-05:00  171.482261  173.286199  170.511657   
237  2021-12-10 00:00:00-05:00  171.776405  176.109784  171.266591   
238  2021-12-13 00:00:00-05:00  177.570578  178.560795  172.090129   
239  2021-12-14 00:00:00-05:00  171.815619  174.256828  168.835201   

          Close     Volume  Dividends  Stock Splits  
235  171.648941  116998900        0.0           0.0  
236  171.139114  108923700        0.0           0.0  
237  175.933304  115402700        0.0           0.0  
238  172.296021  153237000        0.0           0.0  
239  170.913651  139380400        0.0           0.0  
###########################################################################
##################################

## Task 2.4: Boolean Filtering - Multiple Conditions

Find all days where:
1. Close price > $150 AND Volume > 80 million (use `&`)

2. Close price > $180 OR Volume > 120 million (use `|`)

3. Close price is between $140 and $160 (use `.between()`)

For each, display the count and first 5 rows.

In [114]:
# df[(df['Close'] > 150) & (df['Volume'] > 80_000_000)]

## Task 2.5: iloc - Position-based Selection

Using `.iloc`, extract:
1. The value in row 100, column 4 (Close price)
2. Rows 200-210, all columns
3. All rows, columns 1-4 (Open through Close) - display first 10 rows

In [120]:
df.iloc[200:211,:]

,Date,Open,High,Low,Close,Volume,Dividends,Stock Splits
200,2021-10-19 00:00:00-04:00,143.918997,146.033585,143.468677,145.632202,76378900,0.0,0.0
201,2021-10-20 00:00:00-04:00,145.573450,146.601376,145.005644,146.121674,58418800,0.0,0.0
202,2021-10-21 00:00:00-04:00,145.681125,146.493676,144.760887,146.337036,61421000,0.0,0.0
203,2021-10-22 00:00:00-04:00,146.542649,147.022337,145.514723,145.563675,58883400,0.0,0.0
204,2021-10-25 00:00:00-04:00,145.553892,146.229387,144.516182,145.514740,50720600,0.0,0.0
205,2021-10-26 00:00:00-04:00,146.190189,147.668434,145.876910,146.180405,60893400,0.0,0.0
206,2021-10-27 00:00:00-04:00,146.219578,146.581793,145.367875,145.720306,56094900,0.0,0.0
207,2021-10-28 00:00:00-04:00,146.669927,149.949482,146.572024,149.362106,100077900,0.0,0.0
208,2021-10-29 00:00:00-04:00,144.124590,146.787401,143.331623,146.650345,124953200,0.0,0.0
209,2021-11-01 00:00:00-04:00,145.857356,146.552419,144.692374,145.827988,74588300,0.0,0.0


# Exercise 3: Data Transformations

## Task 3.1: Basic Calculations

Create a copy of the DataFrame called `df_work` and add the following columns:

1. `Daily_Range` = High - Low
2. `Range_Pct` = (Daily_Range / Open) * 100
3. `Intraday_Change` = Close - Open
4. `Intraday_Change_Pct` = (Intraday_Change / Open) * 100

Display the first 10 rows with Date, Open, High, Low, Close, and all new columns.

## Task 3.2: Returns Analysis

Calculate the following return metrics:

1. `Returns` = Daily simple returns using `.pct_change()`
2. `Returns_Pct` = Returns in percentage
3. `Cumulative_Returns` = Cumulative returns over the entire period



## Task 3.3: Moving Averages
Calculate the following moving averages on the Close price:

1. `SMA_10` = 10-day simple moving average
2. `SMA_30` = 30-day simple moving average
3. `SMA_50` = 50-day simple moving average

Then:
- Display the last 15 rows showing Date, Close, and all three SMAs
- Create a column `Above_SMA30` that shows 'Above' if Close > SMA_30, else 'Below'
- Count how many days the close was above vs below the 30-day SMA

## Task 3.4: Advanced Features 

Create the following features:

1. `Close_Lag1` = Previous day's close price (use `.shift(1)`)
2. `Close_Lag5` = Close price from 5 days ago
3. `Change_From_5Days_Ago` = Close - Close_Lag5
4. `Pct_Change_From_5Days_Ago` = Percentage change from 5 days ago

Display rows 10-20 with Date, Close, and all lag features.

# Exercise 4: Grouping and Aggregation 

## Task 4.1: Date Feature Extraction

Extract the following from the Date column:
1. `Year`
2. `Month`
3. `Quarter`
4. `Day_of_Week` (0=Monday, 6=Sunday)

Display the first 15 rows with Date and all extracted features.

## Task 4.2: Annual Analysis 

Group the data by Year and calculate:

1. Mean, Median, Min, and Max of the Close price for each year
2. Mean daily return for each year (in %)

Present your results in a well-formatted table.

## Task 4.3: Quarterly Performance

For each quarter in each year:
1. Calculate the mean closing price
2. Calculate the total return for that quarter
3. Identify which quarter had the best performance
4. Identify which quarter had the worst performance

Create a pivot table or grouped table showing Year x Quarter.

## Task 4.4: Day of Week Analysis 

Analyze returns by day of the week:
1. Calculate mean return for each day of the week
2. Calculate the percentage of positive return days for each weekday
3. Determine which day of the week had the highest average return
4. Determine which day had the most volatile returns (highest std)

Present the results in a clear table.